File: unity/ProfitableCore/Simulation/PlanetResourceCycle.cs
using Profitable.Core.Constants;
using Profitable.Core.Schema;

namespace Profitable.Core.Simulation;

// Ports src/galaxy/planetResourceCycle.ts.
public static class PlanetResourceCycle
{
    private const long MsPerHour = 60 * 60 * 1000;

    // Reuses GetCurrentSeason()'s exact technique (src/trading/season.ts,
    // not yet ported -- Sub-Phase B): a seeded phase offset derived from
    // the planet's own id (so planets don't all reset in lockstep),
    // floor-divided into a cycle index. Zero persisted state -- a pure
    // function of (planetId, now).
    public static int GetCycleIndex(string planetId, long nowMs)
    {
        var random = SeededRandom.Create($"{planetId}:resource-cycle-phase");
        var offsetHours = (long)Math.Floor(random() * PlanetResourceCycleConstants.PlanetResourceResetIntervalHours);
        var offsetMs = offsetHours * MsPerHour;
        var intervalMs = (long)PlanetResourceCycleConstants.PlanetResourceResetIntervalHours * MsPerHour;
        return (int)Math.Floor((nowMs + offsetMs) / (double)intervalMs);
    }

    public sealed class ResourcesForCycle
    {
        public List<string> ProducibleResourceIds { get; init; } = new();
        public string? SpecialtyResourceId { get; init; }
        public Dictionary<string, QualityMap> ResourceQualities { get; init; } = new();
    }

    // Factors the subset-selection logic plus the fixed-quality-rolling
    // step into one shared, cycle-parameterized function -- PlanetGenerator
    // calls this with cycleIndex 0 for a planet's initial snapshot, and
    // GetCurrentPlanetResources (below) calls it again with whatever cycle
    // is currently live. One implementation either way, never two copies.
    // Tier, type, and position are NOT re-rolled here -- rolled once on the
    // base seed stream by PlanetGenerator and passed in, permanent for the
    // planet's whole lifetime.
    public static ResourcesForCycle GenerateResourcesForCycle(
        string seed,
        TierColor tier,
        PlanetType planetType,
        IReadOnlyList<Resource> resources,
        int cycleIndex)
    {
        var random = SeededRandom.Create($"{seed}:resources:{cycleIndex}");
        var eligibleResources = ResourceSubsetSelector.GetEligibleResources(planetType, resources);
        if (eligibleResources.Count == 0)
        {
            throw new InvalidOperationException($"no eligible resources for planet type {planetType} in the given catalog");
        }
        var count = ResourceSubsetSelector.ComputeSubsetCount(tier, eligibleResources.Count);
        var selection = ResourceSubsetSelector.SelectResourceSubset(eligibleResources, tier, count, random);

        var resourceQualities = new Dictionary<string, QualityMap>();
        foreach (var id in selection.ProducibleResourceIds)
        {
            var resource = resources.FirstOrDefault(r => r.Id == id);
            if (resource is null) continue;
            resourceQualities[id] = PlanetQualityRoller.RollQualityOnPlanet(resource, tier, selection.SpecialtyResourceId, random);
        }

        return new ResourcesForCycle
        {
            ProducibleResourceIds = selection.ProducibleResourceIds,
            SpecialtyResourceId = selection.SpecialtyResourceId,
            ResourceQualities = resourceQualities,
        };
    }

    // Starting-planet tutorial guarantee. Direct, closed-form override --
    // never a retry-until-valid loop. Reapplied every cycle, including
    // every reset: idempotent, never a one-time bootstrap that could drift
    // away after a reset. Takes its own independent random stream rather
    // than continuing GenerateResourcesForCycle's already-closed-over one.
    private static ResourcesForCycle ApplyTutorialGuarantee(
        ResourcesForCycle baseResources,
        IReadOnlyList<Resource> resources,
        TierColor tier,
        RandomFn random)
    {
        var producibleResourceIds = new List<string>(baseResources.ProducibleResourceIds);
        var resourceQualities = new Dictionary<string, QualityMap>(baseResources.ResourceQualities);

        foreach (var guaranteedId in PlanetResourceCycleConstants.TutorialGuaranteedResourceIds)
        {
            var resource = resources.FirstOrDefault(r => r.Id == guaranteedId);
            if (resource is null) continue; // not in the catalog at all -- nothing to guarantee

            if (!producibleResourceIds.Contains(guaranteedId))
            {
                // Bypasses the normal subset draw for just this slot --
                // added on top of whatever the natural roll produced,
                // never displacing another resource to make room.
                producibleResourceIds.Add(guaranteedId);
                resourceQualities[guaranteedId] = PlanetQualityRoller.RollQualityOnPlanet(
                    resource, tier, baseResources.SpecialtyResourceId, random);
            }

            var roll = resourceQualities[guaranteedId];
            var aggregateTier = AggregateTierResolver.ComputeAggregateTier(roll);
            if (aggregateTier != TierColor.Grey && aggregateTier != TierColor.White)
            {
                var clamped = new QualityMap();
                foreach (var quality in Qualities.All)
                {
                    var value = roll.TryGetValue(quality, out var v) ? v : null;
                    clamped[quality] = value is null ? (int?)null : PlanetResourceCycleConstants.TutorialGuaranteeQualityClamp;
                }
                resourceQualities[guaranteedId] = clamped;
            }
        }

        return new ResourcesForCycle
        {
            ProducibleResourceIds = producibleResourceIds,
            SpecialtyResourceId = baseResources.SpecialtyResourceId,
            ResourceQualities = resourceQualities,
        };
    }

    // The live read path every gameplay caller switches to using INSTEAD OF
    // reading Planet's static fields directly -- those fields on a cached
    // Planet object remain the cycle-0 snapshot only.
    //
    // Colonist-Driven Production (Sub-Phase E, planet-ownership.md): the
    // first check, before anything else runs. Planet.ColonistCount is null
    // for a raw, unmerged Planet -- treated as 0, same as an explicit 0, so
    // an unmerged Planet always reads as ungatherable rather than silently
    // bypassing the gate.
    public static ResourcesForCycle GetCurrentPlanetResources(
        Planet planet,
        IReadOnlyList<Resource> resources,
        long nowMs,
        bool isStartingPlanet = false)
    {
        if ((planet.ColonistCount ?? 0) < PlanetOwnershipConstants.MinimumColonistsToProduce)
        {
            return new ResourcesForCycle();
        }

        if (planet.Tier is null || planet.PlanetType is null)
        {
            throw new InvalidOperationException($"GetCurrentPlanetResources: planet {planet.Id} is missing Tier/PlanetType");
        }

        var cycleIndex = GetCycleIndex(planet.Id, nowMs);
        var baseResources = GenerateResourcesForCycle(planet.Id, planet.Tier.Value, planet.PlanetType.Value, resources, cycleIndex);

        if (!isStartingPlanet)
        {
            return baseResources;
        }

        var random = SeededRandom.Create($"{planet.Id}:resources:{cycleIndex}:tutorial-guarantee");
        return ApplyTutorialGuarantee(baseResources, resources, planet.Tier.Value, random);
    }
}

File: src/galaxy/planetResourceCycle.ts
import type { Resource } from "../data/types/resource.ts";
import type { Planet } from "../data/types/planet.ts";
import type { PlanetType } from "../data/types/planetType.ts";
import type { TierColor } from "../data/types/tierColor.ts";
import type { QualityRoll } from "../data/types/quality.ts";
import type { RandomFn } from "../data/types/random.ts";
import { QUALITIES } from "../data/types/quality.ts";
import {
  PLANET_RESOURCE_RESET_INTERVAL_HOURS,
  TUTORIAL_GUARANTEED_RESOURCE_IDS,
  TUTORIAL_GUARANTEE_QUALITY_CLAMP,
} from "../data/constants/planetResourceCycle.ts";
import { MINIMUM_COLONISTS_TO_PRODUCE } from "../data/constants/planetOwnership.ts";
import { createSeededRandom } from "./seededRandom.ts";
import { getEligibleResources, computeSubsetCount, selectResourceSubset } from "./resourceSubset.ts";
import { rollQualityOnPlanet } from "./rollQualityOnPlanet.ts";
import { computeAggregateTier } from "../simulation/aggregateTier.ts";

const MS_PER_HOUR = 60 * 60 * 1000;

// Planet Resource Generation (profitable-design-questions.md). Reuses
// getCurrentSeason()'s exact technique (src/trading/season.ts): a seeded
// phase offset derived from the planet's own id (so planets don't all
// reset in lockstep), floor-divided into a cycle index. Zero persisted
// state -- a pure function of (planetId, now), same "always live, never
// stale" property galaxy.md already guarantees extended to one more layer.
export function getPlanetResourceCycleIndex(planetId: string, now: number): number {
  const random = createSeededRandom(`${planetId}:resource-cycle-phase`);
  const offsetHours = Math.floor(random() * PLANET_RESOURCE_RESET_INTERVAL_HOURS);
  const offsetMs = offsetHours * MS_PER_HOUR;
  const intervalMs = PLANET_RESOURCE_RESET_INTERVAL_HOURS * MS_PER_HOUR;
  return Math.floor((now + offsetMs) / intervalMs);
}

export interface ResourcesForCycle {
  producibleResourceIds: string[];
  specialtyResourceId: string | null;
  resourceQualities: Record<string, QualityRoll>;
}

// Factors the existing subset-selection logic (generatePlanet.ts) plus the
// new fixed-quality-rolling step out into one shared, cycle-parameterized
// function -- generatePlanet() calls this with cycleIndex 0 for a planet's
// initial snapshot, and getCurrentPlanetResources() (below) calls it again
// with whatever cycle is currently live. One implementation either way,
// never two copies of the same formula. Tier, type, and position are NOT
// re-rolled here -- rolled once on the base seed stream by generatePlanet()
// and passed in, permanent for the planet's whole lifetime.
export function generateResourcesForCycle(
  seed: string,
  tier: TierColor,
  planetType: PlanetType,
  resources: Resource[],
  cycleIndex: number,
): ResourcesForCycle {
  const random = createSeededRandom(`${seed}:resources:${cycleIndex}`);
  const eligibleResources = getEligibleResources(planetType, resources);
  if (eligibleResources.length === 0) {
    throw new Error(`no eligible resources for planet type ${planetType} in the given catalog`);
  }
  const count = computeSubsetCount(tier, eligibleResources.length);
  const { producibleResourceIds, specialtyResourceId } = selectResourceSubset(
    eligibleResources,
    tier,
    count,
    random,
  );

  const resourceQualities: Record<string, QualityRoll> = {};
  for (const id of producibleResourceIds) {
    const resource = resources.find((entry) => entry.id === id);
    if (!resource) continue;
    resourceQualities[id] = rollQualityOnPlanet(resource, { tier, specialtyResourceId }, random);
  }

  return { producibleResourceIds, specialtyResourceId, resourceQualities };
}

// Starting-planet tutorial guarantee. Direct, closed-form override -- never
// a retry-until-valid loop (this codebase has never used one). Reapplied
// every cycle, including every reset: idempotent, never a one-time
// bootstrap that could drift away after a reset. Takes its own random
// stream (a separate seed suffix, below) rather than continuing
// generateResourcesForCycle()'s already-closed-over one -- the same
// "independent seeded stream per concern" pattern galaxy.md's
// position-vs-index streams and season.ts's phase-vs-effect streams
// already use, not a new discipline invented here.
function applyTutorialGuarantee(
  base: ResourcesForCycle,
  resources: Resource[],
  tier: TierColor,
  random: RandomFn,
): ResourcesForCycle {
  let producibleResourceIds = base.producibleResourceIds;
  const resourceQualities = { ...base.resourceQualities };

  for (const guaranteedId of TUTORIAL_GUARANTEED_RESOURCE_IDS) {
    const resource = resources.find((entry) => entry.id === guaranteedId);
    if (!resource) continue; // not in the catalog at all -- nothing to guarantee

    if (!producibleResourceIds.includes(guaranteedId)) {
      // Bypasses the normal subset draw for just this slot -- added on top
      // of whatever the natural roll produced, never displacing another
      // resource to make room.
      producibleResourceIds = [...producibleResourceIds, guaranteedId];
      resourceQualities[guaranteedId] = rollQualityOnPlanet(
        resource,
        { tier, specialtyResourceId: base.specialtyResourceId },
        random,
      );
    }

    const roll = resourceQualities[guaranteedId]!;
    const aggregateTier = computeAggregateTier(roll);
    if (aggregateTier !== "Grey" && aggregateTier !== "White") {
      const clamped = {} as QualityRoll;
      for (const quality of QUALITIES) {
        clamped[quality] = roll[quality] === null ? null : TUTORIAL_GUARANTEE_QUALITY_CLAMP;
      }
      resourceQualities[guaranteedId] = clamped;
    }
  }

  return { producibleResourceIds, specialtyResourceId: base.specialtyResourceId, resourceQualities };
}

// The live read path every gameplay caller switches to using INSTEAD OF
// reading planet.producibleResourceIds/specialtyResourceId/resourceQualities
// directly -- those fields on a cached Planet object remain the cycle-0
// snapshot only, useful for "as originally generated," but no longer the
// authoritative "what can be mined right now" answer once a planet has
// lived past its first reset interval.
//
// Necessary completion: the design entry's own signature named only
// (planet, now, isStartingPlanet?) -- structurally incomplete, since
// regenerating a cycle's resource subset needs the resource catalog to
// resolve ids into real Resource objects, the same "resources" parameter
// generatePlanet()/generateResourcesForCycle() already require. Added here
// as a required parameter, the same class of necessary completion as the
// MVP's loadContent()/RefiningRecipe additions.
//
// Colonist-Driven Production (planet-ownership.md): the first check, before
// anything else runs. planet.colonistCount is undefined for a raw,
// unmerged Planet (never merged with planetOwnershipState) -- treated as 0,
// same as an explicit 0, so an unmerged Planet always reads as ungatherable
// rather than silently bypassing the gate. Callers MUST pass a merged
// Planet (mergePlanetOwnership()) for a colonized planet to correctly read
// as productive.
export function getCurrentPlanetResources(
  planet: Planet,
  resources: Resource[],
  now: number,
  isStartingPlanet = false,
): ResourcesForCycle {
  if ((planet.colonistCount ?? 0) < MINIMUM_COLONISTS_TO_PRODUCE) {
    return { producibleResourceIds: [], specialtyResourceId: null, resourceQualities: {} };
  }

  if (!planet.tier || !planet.planetType) {
    throw new Error(`getCurrentPlanetResources: planet ${planet.id} is missing tier/planetType`);
  }

  const cycleIndex = getPlanetResourceCycleIndex(planet.id, now);
  const base = generateResourcesForCycle(planet.id, planet.tier, planet.planetType, resources, cycleIndex);

  if (!isStartingPlanet) {
    return base;
  }

  const random = createSeededRandom(`${planet.id}:resources:${cycleIndex}:tutorial-guarantee`);
  return applyTutorialGuarantee(base, resources, planet.tier, random);
}
